src/main/scala/org/constellation/snapshotstreaming/FileBasedLastGlobalSnapshotStorage.scala

```diff
@@ -1,84 +1,94 @@
+import org.tessellation.schema.transaction.DAGTransaction
+import org.tessellation.schema.block.DAGBlock
+import org.tessellation.schema.GlobalSnapshotInfo
+import cats.data.OptionT
+import org.tessellation.sdk.domain.cluster.storage.L0ClusterStorage
 import org.tessellation.security.Hashed
+import cats.data.OptionT
+import cats.effect.Async
+import cats.syntax.either._
+import cats.syntax.functor._
+import cats.syntax.option._
+
+import org.tessellation.kryo.KryoSerializer
+import org.tessellation.schema.{GlobalSnapshot, IncrementalGlobalSnapshot}
+import org.tessellation.sdk.infrastructure.snapshot.storage.SnapshotLocalFileSystemStorage
+import org.tessellation.security.hash.Hash
+import org.tessellation.security.signature.Signed
 import fs2.io.file._
 import fs2.{Stream, text}
 import io.circe.Codec
 import io.circe.generic.semiauto.deriveCodec
 import io.circe.parser.decode
 import io.circe.syntax._
-import org.typelevel.log4cats.slf4j.Slf4jLogger
-
-object FileBasedLastGlobalSnapshotStorage {
-
-  def make[F[_]: Async: Files](path: Path): F[LastGlobalSnapshotStorage[F]] =
-    Semaphore[F](1)
-      .map(make(path, _))
+import cats.syntax.applicativeError._
+import cats.syntax.semigroupk._
 
-  private def make[F[_]: Async: Files](path: Path, semaphore: Semaphore[F]): LastGlobalSnapshotStorage[F] =
-    new LastGlobalSnapshotStorage[F] {
+import fs2.io.file.Path
+import cats.Applicative
+import org.typelevel.log4cats.slf4j.Slf4jLogger
 
-      implicit val codec: Codec[Hashed[GlobalSnapshot]] = deriveCodec[Hashed[GlobalSnapshot]]
+trait FileBasedLastGlobalSnapshotStorage[F[_]] {
+  def getGlobalSnapshot: F[Option[Signed[GlobalSnapshot]]]
+}
 
-      def set(snapshot: Hashed[GlobalSnapshot]): F[Unit] =
-        semaphore.permit.use { _ =>
-          get.flatMap {
-            case Some(last) if isNextSnapshot(last, snapshot) =>
-              Stream
-                .emit(snapshot.asJson.spaces2)
-                .through(text.utf8.encode)
-                .through(Files[F].writeAll(path))
-                .compile.drain
-            case Some(last) =>
-              MonadThrow[F].raiseError[Unit](new Throwable(s"Snapshot is not the next one! last: ${getSnapshotReference(last)} next: ${getSnapshotReference(snapshot)}"))
-            case None =>
-              MonadThrow[F].raiseError[Unit](new Throwable("Previous snapshot not found when setting next global snapshot!"))
-          }
-        }
+object FileBasedLastGlobalSnapshotStorage {
 
-      def setInitial(snapshot: Hashed[GlobalSnapshot]): F[Unit] =
-        semaphore.permit.use { _ =>
-          Stream
-            .emit(snapshot.asJson.spaces2)
-            .through(text.utf8.encode)
-            .through(Files[F].writeAll(path, Flags(Flag.Write, Flag.CreateNew)))
-            .compile.drain
-        }
+  def make[F[_]: Async](path: Path) = new FileBasedLastGlobalSnapshotStorage[F] {
+    implicit val codec: Codec[Hashed[GlobalSnapshot]] = deriveCodec[Hashed[GlobalSnapshot]]
```

Comment:
I guess if you are not saving or reading the `Hashed[GlobalSnapshot]` you also don't need a codec for it.